### PR TITLE
mui-datatable v3 release

### PR DESCRIPTION
**Pull request by patorjk**
This is currently a work in progress. This release will represent the 3.0.0 version of the table. 

Goals:
* Update all of the dependencies.
* Fully move the table to supporting version 4 of Material UI. 
* Incorporate the many PRs that have been submitted.

It can be previewed here: https://codesandbox.io/s/github/gregnb/mui-datatables/tree/v3

Current Updates:
* Dependencies updated (replaces https://github.com/gregnb/mui-datatables/pull/1077)
* Fixed issues with "responsive" option, added tableBodyHeight and tableBodyMaxHeight options, added new responsive mode "simple" (https://github.com/gregnb/mui-datatables/pull/1312)
* fixedHeaderOptions has been deprecated in favor of fixedHeader and fixedSelectColumn options. Bugs seen in Safari related to this feature have been fixed (https://github.com/gregnb/mui-datatables/pull/1088, https://github.com/gregnb/mui-datatables/issues/900)
* A boolean option for a header to the expand or collapse all rows has been added - expandableRowsHeader (https://github.com/gregnb/mui-datatables/pull/1091)
* A boolean option to hide the checkboxes that appear when rows are selectable has been added - selectableRowsHideCheckboxes (https://github.com/gregnb/mui-datatables/pull/704, https://github.com/gregnb/mui-datatables/pull/726)
* A button to close the filter dialog has been added (https://github.com/gregnb/mui-datatables/pull/1171)
* Bug fix for columns resizing (https://github.com/gregnb/mui-datatables/pull/1172)
* Bug fix for table resizing with resizable columns (https://github.com/gregnb/mui-datatables/pull/1177)
* Changed first time column sort order to be ascending (https://github.com/gregnb/mui-datatables/pull/1274)
* Prevent search icon focus on close (https://github.com/gregnb/mui-datatables/pull/1207)
* Various code cleanup and documentation updates (https://github.com/gregnb/mui-datatables/pull/1206)
* A new prop called "components" has been added that allows developers to override many of the table's internal components. This allows much more customization. The feature has some overlap with customRender functions, but it provides a more friendly API and allows for overrides that aren't currently possible (https://github.com/gregnb/mui-datatables/pull/1167)
* enableNestedDataAccess option added. This option allows users use nested data for column values. (https://github.com/gregnb/mui-datatables/pull/1181, https://github.com/gregnb/mui-datatables/issues/909)
* The onColumnSortChange callback will now receive values of "asc" or "desc" for the sortDirection value, rather than the full names spelled out (https://github.com/gregnb/mui-datatables/pull/995)
* filterData is now propagated to custom filters (https://github.com/gregnb/mui-datatables/pull/1136)
* Prevent skipping of heading level for Accessibility and standard semantic HTML purposes (https://github.com/gregnb/mui-datatables/pull/1154)
* Bug fix for table header checkbox being selected when the table's contents are empty (https://github.com/gregnb/mui-datatables/pull/1304)
* Added filterOptions.renderValue option (https://github.com/gregnb/mui-datatables/pull/1121/)
* Renamed onColumnViewChange to onViewColumnsChange (https://github.com/gregnb/mui-datatables/pull/1160)
* Added searchProps option for adding custom props to the search box (https://github.com/gregnb/mui-datatables/pull/808)
* Added selectToolbarPlacement option: Controls the visibility of the Select Toolbar, options are 'replace' (select toolbar replaces default toolbar when a row is selected), 'above' (select toolbar will appear above default toolbar when a row is selected) and 'none' (select toolbar will never appear). Deprecates the "disableToolbarSelect" option (https://github.com/gregnb/mui-datatables/pull/602)
* Updated the we pack config (https://github.com/gregnb/mui-datatables/pull/1306)
* Added sortOrder option (deprecating sortDirection on columns object) (https://github.com/gregnb/mui-datatables/pull/1310/)
* Toolbar, select cells, and footer now hidden when the table is printed (https://github.com/gregnb/mui-datatables/pull/1269)
* onRowsSelect is now onRowSelectionChange
* onRowsExpand is now onRowExpansionChange
* Select cells, table header, and table footer no longer show when printing the table (https://github.com/gregnb/mui-datatables/pull/1269)
* Added confirmFilters option (deprecates serverSideFilterList)  (https://github.com/gregnb/mui-datatables/pull/1317)
* Added customTableBodyFooterRender (https://github.com/gregnb/mui-datatables/pull/1314)
* Bug fix: Deprecation warnings stopped working starting in version 2.14.0 of the table. They have been re-enabled for version 3.0.0. (https://github.com/gregnb/mui-datatables/issues/1319)

This list will grow over the next week or two. The goal here is to bring in features that have been waiting around a while and to get the table in-line with Material UI v4.

Breaking updates:
* fixedHeaderOptions deprecated.
* Nested data access through the DOT notation in column names is now OFF by default. The enableNestedDataAccess option can be used to turn it on (ex: enableNestedDataAccess: "." would make things work like before).
* The onColumnSortChange callback will now receive values of "asc" or "desc" for the sortDirection value, rather than the full names spelled out (https://github.com/gregnb/mui-datatables/pull/995)
* onColumnViewChange option is renamed to onViewColumnsChange. Corresponding table action columnViewChange renamed to viewColumnsChange (https://github.com/gregnb/mui-datatables/pull/1160)
* disableToolbarSelect deprecated in favor of selectToolbarPlacement.

Existing PRs are being merged into this branch (v3). If you're interested in submitting a PR anytime soon, please checkout v3 and work from that.

### CODE REVIEW

**Comment by wdh2100** [comment]
@patorjk 
Would you like to update the version of react(16.8^)?

**Comment by patorjk** [comment]
Yes, it’s being updated so hooks can be used, though right nothing is yet utilizing hooks.  

I think the peer dependency is a little high and I’ll probably adjust it to 16.8 or 16.9. I’m looking at other libs to see what the best value is. I’m open to suggestions so all thoughts are welcome here!

File: README.md [hunk 1]
```diff
@@ -399,6 +399,9 @@ Supported customizable components:
  * `TableResize`
  * `TableToolbar`
  * `TableToolbarSelect`
+ * `Tooltip`
+
+For more information, please see this [example](https://github.com/gregnb/mui-datatables/examples/custom-components/index.js). Additionally, all examples can be viewd [live](https://codesandbox.io/s/github/greg/mui-datatables) at our CodeSandbox.
```
Comment (wdh2100) [inline README.md:404]: >https://codesandbox.io/s/github/greg/mui-datatables

Could not find the specified repository or directory

Comment (patorjk) [reply README.md]: Good catch! Thank you for checking that - I've updated the README, please let me know if you see anything else like that. I looked into editing the description and it appears only project owners can edit it (I'm just a maintainer). I may hit greg up later about it but I don't want to bug him too much (just asked him a question earlier today so I'll wait a bit).

Comment (wdh2100) [reply README.md]: @patorjk Thanks~

File: src/components/TableBody.js [hunk 2]
```diff
@@ -283,7 +283,7 @@ class TableBody extends React.Component {
               colIndex={0}
               rowIndex={0}>
               <Typography variant="body1" className={classes.emptyTitle}>
-                <strong>{options.textLabels.body.noMatch}</strong>
+                {options.textLabels.body.noMatch}
```
Comment (wdh2100) [inline src/components/TableBody.js:286]: Is there a reason for removing the strong(bold)?

It's different.
https://codesandbox.io/s/material-demo-buwx0?file=/demo.js

Comment (wdh2100) [reply src/components/TableBody.js:286]: oh.. https://github.com/gregnb/mui-datatables/pull/1154

I found it here.

that's right 
It seems right to remove strong tag.

**Comment by kev-landry** [comment]
> Yes, it’s being updated so hooks can be used, though right nothing is yet utilizing hooks.

So that means we could use hooks like useMemo to avoid to perform the sorting function twice on the same column ?

**Comment by patorjk** [comment]
> > Yes, it’s being updated so hooks can be used, though right nothing is yet utilizing hooks.
> 
> So that means we could use hooks like useMemo to avoid to perform the sorting function twice on the same column ?

Yes. and I'd definitely be open any performance optimizations like this. However, using useMemo in some of table's internal components will require some refactoring and I think I want to wait until after the 3.0.0 release for that to happen (since there's still a dozen or so PRs to work through and I want to make sure everything is tested and working correctly).

**Comment by patorjk** [comment]
I've released a beta version to npm which can be installed like this:

    npm install mui-datatables@beta

I still need to do some more testing, but I'm thinking 3.0.0 will be released either tomorrow or Wednesday.

**Comment by kev-landry** [comment]
Well done :)

**Pull request merged**